drivers/SmartThings/matter-switch/src/switch_handlers/attribute_handlers.lua

```diff
@@ -267,45 +270,82 @@ end
+    if ib.data.elements.energy then
+      local energy_imported_wh = ib.data.elements.energy.value / 1000 -- convert received milliwatt-hour to watt-hour
+      if is_periodic_report then
+        -- handle this report only if cumulative reports are not supported
+        if device:get_field(fields.CUMULATIVE_REPORTS_SUPPORTED) then return end
+        local state_device = switch_utils.find_child(device, ib.endpoint_id) or device
+        local energy_meter_latest_state = state_device:get_latest_state(
+          "main", capabilities.energyMeter.ID, capabilities.energyMeter.energy.NAME, 0 -- 0 as the default if state is nil
+        )
+        energy_imported_wh = energy_imported_wh + energy_meter_latest_state
+      end
+      device:emit_event_for_endpoint(ib.endpoint_id, capabilities.energyMeter.energy({ value = energy_imported_wh, unit = "Wh" }))
+      switch_utils.report_power_consumption_to_st_energy(device, ib.endpoint_id, energy_imported_wh)
     else
-      -- when energy management is defined in the root endpoint(0), replace it with the first switch endpoint and process it.
-      device:emit_event_for_endpoint(device:get_field(fields.ENERGY_MANAGEMENT_ENDPOINT), capabilities.energyMeter.energy({ value = watt_hour_value, unit = "Wh" }))
+      device.log.warn("Received data from the energy imported attribute does not include a numerical energy value")
     end
-    switch_utils.report_power_consumption_to_st_energy(device, device:get_field(fields.TOTAL_IMPORTED_ENERGY))
   end
 end
 
-function AttributeHandlers.per_energy_imported_handler(driver, device, ib, response)
-  if ib.data.elements.energy then
-    local watt_hour_value = ib.data.elements.energy.value / fields.CONVERSION_CONST_MILLIWATT_TO_WATT
-    local latest_energy_report = device:get_field(fields.TOTAL_IMPORTED_ENERGY) or 0
-    local summed_energy_report = latest_energy_report + watt_hour_value
-    device:set_field(fields.TOTAL_IMPORTED_ENERGY, summed_energy_report, {persist = true})
-    device:emit_event(capabilities.energyMeter.energy({ value = summed_energy_report, unit = "Wh" }))
-    switch_utils.report_power_consumption_to_st_energy(device, device:get_field(fields.TOTAL_IMPORTED_ENERGY))
+
+-- [[ POWER TOPOLOGY CLUSTER ATTRIBUTES ]] --
+
+--- AvailableEndpoints: This attribute SHALL indicate the list of endpoints capable of
+--- providing power to and/or consuming power from the endpoint hosting this server.
+---
+--- In the case there are multiple endpoints supporting the PowerTopology cluster with
+--- SET feature, all AvailableEndpoints responses must be handled before profiling.
+function AttributeHandlers.available_endpoints_handler(driver, device, ib, response)
+  local set_topology_eps = device:get_field(fields.ELECTRICAL_SENSOR_EPS)
+  for i, set_ep_info in pairs(set_topology_eps or {}) do
+    if ib.endpoint_id == set_ep_info.endpoint_id then
+      switch_utils.remove_field_index(device, fields.ELECTRICAL_SENSOR_EPS, i) -- ep found, remove from table
```

Comment:
I might just be missing context, but for both of these handlers (`available_endpoint_handler` and `parts_list_handler`) I don't understand what is being done by this function and why this endpoint is being removed. I think some additional documentation could be useful here for developers outside this scope of expertise. Maybe just a quick line explaining why these endpoints are being processed and why the source endpoint for this handler seems to be removed from the list that is later processed.